[MIG] l10n_es_vat_book: Migration to V10 from V11

igallart opened the pull request:
Hola, hemos migrado (downgrade) el módulo de la V11 a la V10 de l10n_es_vat_book. Hemos puesto manualmente los commits siguientes:

https://github.com/OCA/l10n-spain/commit/d4b9b306fc6ad4b744a22f5daeeccd10aa76d85d
https://github.com/OCA/l10n-spain/commit/59cc883d1aee608fb39ad67b31fbe297a0de633f

No tengo claro el procedimiento cuando se trata de pasar un módulo de una versión más alta a una más vieja. Ya vamos mejorándolo entre todos.

carlosdauden committed:
[IMP] l10n_es_aeat: Añadir campo aeat_anonymous_cash_customer y funcion parse_vat_info

carlosdauden committed:
[IMP] l10n_es_aeat: Añadir método que devuelve impuestos desde plantillas

pedrobaeza commented:
Buenas. La idea sería hacer cherry-pick de esos commits resolviendo conflictos.

igallart committed:
[10.0] Migración de l10n_es_vat_book desde 11.0

igallart commented:
> Buenas. La idea sería hacer cherry-pick de esos commits resolviendo conflictos.

Creo que ahora sí.

pedroortega-punt commented:
Buenas.

Realizando la migración del módulo, hemos visto que habría que hacer unos cambios para todas las versiones (V10, V11 y V12).

El libro de IVA no deja realizar la confirmación cuando un cliente anónimo no tiene NIF y cuando un cliente Extracomunitario no tiene NIF, por lo que habría que quitar las restricción para estos dos tipos de clientes.

Por otro lado la renumeración de los registros de las facturas de compras no está funcionando. Para esta renumeración hay dos tipos de empresas las cuales llevan dicha numeración diferente. Está el cliente que quiere llevar el secuenciador de facturas de compra de Odoo sin tener saltos de número y el que no hace caso del número de factura de Odoo porque espera que el libro de IVA lleve la numeración de registros. 

Mi propuesta es la siguiente: 
Cuando se marque el check para renumerar el libro de Iva, se debería programar que fuese obligatorio primero generar el libro de IVA anual, para poder generar cualquier mes o trimestre e ir recalculando para que así el número de registro se mantenga y no se renumere cada vez desde el número 1, ya que hacienda no permite que en el número de registro hayan saltos y deben de ser correlativos.

pedrobaeza commented:
> El libro de IVA no deja realizar la confirmación cuando un cliente anónimo no tiene NIF

El check de cliente anónimo está en este módulo o sus dependencias (me suena que algo incorporo mi compañero @carlosdauden en `l10n_es_aeat` al respecto)? ¿Podéis añadir en un commit aparte esta mejora y lo llevamos a todas las versiones?

> cuando un cliente Extracomunitario no tiene NIF, por lo que habría que quitar las restricción para estos dos tipos de clientes.

No tengo claro que esto deba ser así. @albertgafic  @acysos podéis confirmar? ¿Cómo lo compruebas de todas formas? ¿Por país?

carlosdauden commented:
@vipvalen En v11 y v12 cualquier partner puede marcarse como anónimo para que no se tenga en cuenta:
https://github.com/OCA/l10n-spain/blob/107ae0eab62497b7c7853ff065d7fbc884ac3a0b/l10n_es_vat_book/models/l10n_es_vat_book.py#L329-L331

Respecto a la renumeración, creo que es una opción poco necesaria actualmente, pero cualquier PR de mejora es bienvenido

pedroortega-punt commented:
Ok, y los extracomunitarios que por ejemplo en china no te van a dar un NIF?

pedroortega-punt commented:
Este es el Real Decreto de las facturas recibidas, las cuales obliga a un número de recepción, dado que el libro de Iva lo recoge de apuntes contables y no de facturas recibidas, por lo que debería de ser numerado con un número de registro.

![imagen](https://user-images.githubusercontent.com/18347857/73661042-ab885280-4699-11ea-9c9e-1bf3d0fcaf4a.png)

igallart committed:
[FIX] l10n_es_vat_book: Eliminamos error de aviso de "Sin NIF"

Para clientes/proveedores que son extracomunitarios no añadimos
la excepción de que falta el NIF.

igallart commented:
Hemos hecho una modificación para tener en cuenta el tema del NIF en extracomunitarios. Mirad  a ver que os parece.

pedroortega-punt commented:
@pedrobaeza  Ahora se podría fusionar realizado este commit?.

Pedro que te parece la idea de numeración anual para poder generar los libros de IVA mensuales o trimestrales y que siempre saque el número correcto y no empiece cada vez por el nº1 en recibidas? Si te parece bien crearemos un pull request para ello.

rafaelbn reviewed (approved):
👍 @acysos @jalzaga podéis revisar por favor?

jalzaga commented:
He probado el módulo en Runbot y cuando tienes facturas con recargo de equivalencia da un error al calcular:
`Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 642, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 684, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 334, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 101, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 327, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 942, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 507, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/odoo-10.0/addons/web/controllers/main.py", line 899, in call_button
    action = self._call_kw(model, method, args, {})
  File "/home/odoo/odoo-10.0/addons/web/controllers/main.py", line 887, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 689, in call_kw
    return call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 680, in call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/odoo/build/OCA/l10n-spain/l10n_es_aeat/models/l10n_es_aeat_report.py", line 303, in button_recalculate
    return self.calculate()
  File "/home/odoo/build/OCA/l10n-spain/l10n_es_vat_book/models/l10n_es_vat_book.py", line 178, in calculate
    self._calculate_vat_book()
  File "/home/odoo/build/OCA/l10n-spain/l10n_es_vat_book/models/l10n_es_vat_book.py", line 388, in _calculate_vat_book
    self.create_vat_book_lines(lines_issued, 'issued', taxes_issued)
  File "/home/odoo/build/OCA/l10n-spain/l10n_es_vat_book/models/l10n_es_vat_book.py", line 346, in create_vat_book_lines
    self.upsert_book_line_tax(move_line, moves_dic[line_key], taxes)
  File "/home/odoo/build/OCA/l10n-spain/l10n_es_vat_book/models/l10n_es_vat_book.py", line 266, in upsert_book_line_tax
    lambda l: not l['special_tax_group'], implied_lines))
TypeError: list object is not an iterator`

jalzaga commented:
Por otro lado, cuando entras en los detalles de facturas emitidas o recibidas en cualquier factura, si vas a las líneas de impuestos y pinchas da un error:
`Error: Unknown field parent.special_tax_group in domain [["parent.special_tax_group","=",false]]`

igallart commented:
Gracias @jalzaga , he intentado quitar algunos errores de travis y al hacer de nuevo el push se me ha pasado lo del dominio, repararé tambié lo del recargo en cuanto pueda.

igallart commented:
> Por otro lado, cuando entras en los detalles de facturas emitidas o recibidas en cualquier factura, si vas a las líneas de impuestos y pinchas da un error:
> `Error: Unknown field parent.special_tax_group in domain [["parent.special_tax_group","=",false]]`

Corregido.

> He probado el módulo en Runbot y cuando tienes facturas con recargo de equivalencia da un error al calcular:
> `Traceback (most recent call last): File "/.repo_requirements/odoo/odoo/http.py", line 642, in _handle_exception return super(JsonRequest, self)._handle_exception(exception) File "/.repo_requirements/odoo/odoo/http.py", line 684, in dispatch result = self._call_function(**self.params) File "/.repo_requirements/odoo/odoo/http.py", line 334, in _call_function return checked_call(self.db, *args, **kwargs) File "/.repo_requirements/odoo/odoo/service/model.py", line 101, in wrapper return f(dbname, *args, **kwargs) File "/.repo_requirements/odoo/odoo/http.py", line 327, in checked_call result = self.endpoint(*a, **kw) File "/.repo_requirements/odoo/odoo/http.py", line 942, in __call__ return self.method(*args, **kw) File "/.repo_requirements/odoo/odoo/http.py", line 507, in response_wrap response = f(*args, **kw) File "/home/odoo/odoo-10.0/addons/web/controllers/main.py", line 899, in call_button action = self._call_kw(model, method, args, {}) File "/home/odoo/odoo-10.0/addons/web/controllers/main.py", line 887, in _call_kw return call_kw(request.env[model], method, args, kwargs) File "/.repo_requirements/odoo/odoo/api.py", line 689, in call_kw return call_kw_multi(method, model, args, kwargs) File "/.repo_requirements/odoo/odoo/api.py", line 680, in call_kw_multi result = method(recs, *args, **kwargs) File "/home/odoo/build/OCA/l10n-spain/l10n_es_aeat/models/l10n_es_aeat_report.py", line 303, in button_recalculate return self.calculate() File "/home/odoo/build/OCA/l10n-spain/l10n_es_vat_book/models/l10n_es_vat_book.py", line 178, in calculate self._calculate_vat_book() File "/home/odoo/build/OCA/l10n-spain/l10n_es_vat_book/models/l10n_es_vat_book.py", line 388, in _calculate_vat_book self.create_vat_book_lines(lines_issued, 'issued', taxes_issued) File "/home/odoo/build/OCA/l10n-spain/l10n_es_vat_book/models/l10n_es_vat_book.py", line 346, in create_vat_book_lines self.upsert_book_line_tax(move_line, moves_dic[line_key], taxes) File "/home/odoo/build/OCA/l10n-spain/l10n_es_vat_book/models/l10n_es_vat_book.py", line 266, in upsert_book_line_tax lambda l: not l['special_tax_group'], implied_lines)) TypeError: list object is not an iterator`

Corregido.

jalzaga commented:
He probado el módulo en runbot:
Introducidas facturas con diferentes posiciones fiscales: Régimen nacional, régimen intracomunitario, régimen extracomunitario: OK
Recargo de equivalencia: Probado OK, Marca Special group como R.Eq.
Cliente anónimo: Deja confirmar cuando el cliente tiene el indicador de anónimo y no tiene NIF.
Observaciones:
Facturas con IRPF: No marca Special Group como IRPF 
Facturas rectificativas: No veo que se muestre la factura rectificada.

Revisadas estas últimas observaciones, por mi parte el módulo funciona correctamente.

pedroortega-punt commented:
Travis no relacionado con este módulo

************* Module l10n_es_dua
l10n_es_dua/data/products_dua.xml:4: [W7902(duplicate-xml-record-id), ]  Duplicate xml record id "data/fp_dua_noupdate_0" in data/fiscal_positions_dua.xml:4
************* Module l10n_es_aeat_vat_prorrate_asset.models.account_invoice
l10n_es_aeat_vat_prorrate_asset/models/account_invoice.py:22: [W8104(api-one-deprecated), AccountInvoiceLine.asset_create] api.one deprecated
************* Module l10n_es_aeat_mod347.models.mod347
l10n_es_aeat_mod347/models/mod347.py:180: [W0622(redefined-builtin), L10nEsAeatMod347Report._get_taxes] Redefining built-in 'map'
************* Module l10n_es_aeat_mod347.models.spanish_states_mapping
l10n_es_aeat_mod347/models/spanish_states_mapping.py:1: [C8201(no-utf8-coding-comment), ] No UTF-8 coding comment found: Use `# coding: utf-8` or `# -*- coding: utf-8 -*-`
************* Module l10n_es_aeat_mod347.controllers.main
l10n_es_aeat_mod347/controllers/main.py:1: [C8201(no-utf8-coding-comment), ] No UTF-8 coding comment found: Use `# coding: utf-8` or `# -*- coding: utf-8 -*
-`

Hay algún error que veáis del módulo l10n_es_vat_book en travis que no esté corregido?

pedroortega-punt commented:
@pedrobaeza 
Por favor se puede realizar el Merge? o falta algo por revisar que corresponda al l10n_es_vat_book.

pedrobaeza commented:
Habría que solucionar los problemas de Travis que sí son de este módulo (ignorad los warnings), y además hay comentarios de @jalzaga sin solucionar.

pedroortega-punt commented:
@jalzaga  en las observaciones que comentas:
Cuando indicas "Facturas con IRPF" no marca special group, a que te refieres?, ya que el IRPF de momento no está en el Libro de IVA como impuesto.

En facturas rectificativas: No veo que se muestre la factura rectificada. En la V11y V12 está comentado el código y no sabemos el motivo.

# TODO: Substitute Invoice
# sheet.write('J' + str(row),
#             line.invoice_id.refund_invoice_id.number or '')

pedroortega-punt commented:
@pedrobaeza  Ya está corregidos el travis correspondiente a vat_book.

Lo de Jose Maria son observaciones.

Podemos mergear y ya vemos en nuevos commits lo de la rectificativa para v10, v11 y v12?.

pedrobaeza commented:
Travis sigue en rojo

igallart commented:
> Travis sigue en rojo

@pedrobaeza , sí que sigue en rojo, pero si no estoy equivocado son cosas no relacionadas con este pull request, si no es así y puedes indicarme lo que es lo intento solucionar.

rafamarpe commented:
Buenas tardes.
Como esta este pull request?
Nosotros también lo necesitamos para un cliente que ha recibido un requerimiento, tenemos plazo de entrega.

He mirado travis y no parece que tenga nada que ver con el vat_book.

pedrobaeza commented:
Sí que debe tener que ver, porque la rama principal está verde: https://travis-ci.org/OCA/l10n-spain/branches.

El que error no sea en este módulo no significa que no lo provoque este PR, ya que entre ellos pueden surgir interacciones, así que revisa por favor qué de este PR está provocando el problema

igallart commented:
@pedrobaeza , en el primer test que está en rojo solo veo esto:

`************* Module l10n_es_dua

l10n_es_dua/data/products_dua.xml:4: [W7902(duplicate-xml-record-id), ]  Duplicate xml record id "data/fp_dua_noupdate_0" in data/fiscal_positions_dua.xml:4

************* Module l10n_es_aeat_vat_prorrate_asset.models.account_invoice

l10n_es_aeat_vat_prorrate_asset/models/account_invoice.py:22: [W8104(api-one-deprecated), AccountInvoiceLine.asset_create] api.one deprecated

************* Module l10n_es_aeat_mod347.models.mod347

l10n_es_aeat_mod347/models/mod347.py:180: [W0622(redefined-builtin), L10nEsAeatMod347Report._get_taxes] Redefining built-in 'map'

************* Module l10n_es_aeat_mod347.models.spanish_states_mapping

l10n_es_aeat_mod347/models/spanish_states_mapping.py:1: [C8201(no-utf8-coding-comment), ] No UTF-8 coding comment found: Use `# coding: utf-8` or `# -*- coding: utf-8 -*-`

************* Module l10n_es_aeat_mod347.controllers.main

l10n_es_aeat_mod347/controllers/main.py:1: [C8201(no-utf8-coding-comment), ] No UTF-8 coding comment found: Use `# coding: utf-8` or `# -*- coding: utf-8 -*-`

************* Module l10n_es_aeat

Using config file /home/travis/maintainer-quality-tools/travis/cfg/travis_run_pylint_pr.cfg

t_action.js:79: [W7903(javascript-lint), ] Expected indentation of 4 spaces but found 0. [Error/indent]

************* Module account_balance_reporting_xlsx

account_balance_reporting_xlsx/README.rst:1: [W7908(missing-newline-extrafiles), ]  Missing newline

************* Module l10n_es_aeat_sii.models.aeat_sii_map

l10n_es_aeat_sii/models/aeat_sii_map.py:15: [W8104(api-one-deprecated), AeatSiiMap._unique_date_range] api.one deprecated

************* Module l10n_es_aeat_sii.wizards.aeat_sii_password

l10n_es_aeat_sii/wizards/aeat_sii_password.py:29: [W0622(redefined-builtin), pfx_to_pem] Redefining built-in 'file'

l10n_es_aeat_sii/wizards/aeat_sii_password.py:42: [W0622(redefined-builtin), pfx_to_crt] Redefining built-in 'file'

************* Module account_balance_reporting.models.account_balance_reporting_template

account_balance_reporting/models/account_balance_reporting_template.py:62: [W8106(method-required-super), AccountBalanceReportingTemplate.copy] Missing `super` call in "copy" method.

************* Module l10n_es

l10n_es/data/account_chart_template_post.xml:33: [W7902(duplicate-xml-record-id), ]  Duplicate xml record id "data/account_chart_template_assoc_noupdate_0" in data/account_chart_template.xml:34

l10n_es/data/account_chart_template_post.xml:24: [W7902(duplicate-xml-record-id), ]  Duplicate xml record id "data/account_chart_template_pymes_noupdate_0" in data/account_chart_template.xml:44

l10n_es/data/account_chart_template.xml:16: [W7902(duplicate-xml-record-id), ]  Duplicate xml record id "data/common_transfer_account_id_noupdate_0" in data/account_chart_template.xml:66

l10n_es/data/account_chart_template_post.xml:15: [W7902(duplicate-xml-record-id), ]  Duplicate xml record id "data/account_chart_template_full_noupdate_0" in data/account_chart_template.xml:54

l10n_es/data/account_chart_template_post.xml:6: [W7902(duplicate-xml-record-id), ]  Duplicate xml record id "data/account_chart_template_common_noupdate_0" in data/account_chart_template.xml:25

l10n_es/data/taxes_common.xml:490: [W7907(duplicate-xml-fields), ]  Duplicate xml field "amount_type" in lines 492

************* Module l10n_es_toponyms

l10n_es_toponyms/wizard/l10n_es_toponyms_zipcodes.xml:1: [W7930(file-not-used), ]  not used from manifest

************* Module l10n_es_partner

l10n_es_partner/wizard/data_banks.xml:1: [W7930(file-not-used), ]  not used from manifest

`

y no veo nada que tenga relación con las programaciones de este pull request. ¿Puedes indicarme algún ejemplo en el test 1 de lo que está mal? Vamos un poco perdidos con esto de travis.

pedrobaeza commented:
La línea que da error es:
```
l10n_es_aeat/tests/test_l10n_es_aeat.py:89:1: W391 blank line at end of file
```
lo demás son warnings.

igallart commented:
> La línea que da error es:
> 
> ```
> l10n_es_aeat/tests/test_l10n_es_aeat.py:89:1: W391 blank line at end of file
> ```
> 
> lo demás son warnings.

Gracias, ya está cambiado, veremos como sale ahora travis.

pedrobaeza commented:
Bueno, ése no era el único problema. Ese problema era en la fase de lint. Los tests también están incorrectos.

igallart commented:
He intentado solucionar el travis, pero me aparece esto al hacer lo siguiente:

> with self.assertRaises(exceptions.ValidationError):
> self.partner.vat =  '12345678Z'

AssertionError: ValidationError not raised

He visto varios ejemplos de la OCA donde está igual, no sé que puede estar mal, alguna ayuda?

igallart commented:
@pedrobaeza , travis correcto ;)

pedrobaeza commented:
OK, paso a fusionar aunque no lo he revisado, pero por no demorar más su inclusión. Gracias por el trabajo.

/ocabot merge major

pedrobaeza commented:
/ocabot merge major

igallart commented:
@pedrobaeza , no sabía que ya se subían automáticamente las versiones de los módulos, por tanto, han subido dos niveles, el que subimos nosotros más el del bot OCA, no sé si es algún inconveniente.

pedrobaeza commented:
Sin inconveniente. No te preocupes.

igallart commented:
Vale perfecto, gracias.
